File: video/avbuf.c
/*
gcc -fPIC -shared -o avbuf.so avbuf.c
gcc -Iffmpeg/include -fPIC -shared -o avbuf.dll avbuf.c
*/

#include <stdbool.h>
#include <stdint.h>
#include <math.h>
#include <libavutil/error.h>
#include <libavformat/avformat.h>

typedef struct {
	uint8_t *ptr;
	int64_t offset;
	int64_t size;
} Avbuf;

int Avbuf_read(void *ptr, uint8_t *buf, int len) {
	Avbuf *avbuf = (Avbuf *)ptr;

	if (avbuf->offset + len > avbuf->size)
		len = avbuf->size - avbuf->offset;
	if (len == 0)
		return AVERROR_EOF;

	memcpy(buf, avbuf->ptr + avbuf->offset, len);
	avbuf->offset += len;

	return len;
}

int64_t Avbuf_seek(void *ptr, int64_t pos, int whence) {
	Avbuf *avbuf = (Avbuf *)ptr;

	if (whence == AVSEEK_SIZE)
		return avbuf->size;

	avbuf->offset = pos;

	return pos;
}

File: video/avbuf.lua
local ffi = require("ffi")

local avbuf = ffi.load("aqua/video/avbuf.so")

ffi.cdef [[
	typedef struct {
		uint8_t *ptr;
		int64_t offset;
		int64_t size;
	} Avbuf;
	int Avbuf_read(void *ptr, uint8_t *buf, int len);
	int64_t Avbuf_seek(void *ptr, int64_t pos, int whence);
]]

---@class ffi.namespace*
---@field Avbuf_read fun(ptr: ffi.cdata*, buf: ffi.cdata*, len: integer): integer
---@field Avbuf_seek fun(ptr: ffi.cdata*, pos: integer, whence: integer): integer

return avbuf
